JumpscaleLib/sal_zos/coredns/CoreDns.py

```
@@ -0,0 +1,146 @@
+import time
+import json
+from jumpscale import j
+from .. import templates
+from ..abstracts import Nics, Service
+
+logger = j.logger.get(__name__)
+DEFAULT_PORT = 53
+
+class Coredns(Service):
+    """
+    CoreDNS is a DNS server. It is written in Go
+    """
+
+    def __init__(self, name, node, etcd_endpoint, recursive_resolvers, zt_identity=None, nics=None):
+        super().__init__(name, node, 'coredns', [DEFAULT_PORT])
+        self.name = name
+        self.id = 'coredns.{}'.format(self.name)
+        self.node = node
+        self._container = None
+        self.flist = 'https://hub.grid.tf/tf-official-apps/coredns-1.2.2.flist'
+        self.etcd_endpoint = etcd_endpoint
+        self.node_port = None
+        self.recursive_resolvers = "8.8.8.8:53 1.1.1.1:53"
+
+        self._config_dir = '/usr/bin'
+        self._config_name = 'coredns.conf'
+        self.zt_identity = zt_identity
+        self.nics = Nics(self)
+        if nics:
+            for nic in nics:
+                nicobj = self.nics.add(nic['name'], nic['type'], nic['id'], nic.get('hwaddr'))
+                if nicobj.type == 'zerotier':
+                    nicobj.client_name = nic.get('ztClient')
+        if 'nat0' not in self.nics:
+            self.nics.add('nat0', 'default')
+
+
+    @property
+    def _container_data(self):
+        """
+        :return: data used for coredns container
+         :rtype: dict
+        """
+        self.node_port = DEFAULT_PORT
```

maybe check if the port is not already used on the node, and if it is the case, raise an error
otherwise we are able to deploy multiple containers forwarding the same port, this is going to be a hell to debug

---

but any empty node , it's already use 53 port : 

```{'ip': 'fe80::ec55:5aff:fe74:180d',
  'network': 'udp6',
  'pid': 602,
  'port': 53},
 {'ip': 'fe80::24a9:b3ff:fe9a:8b25',
  'network': 'udp6',
  'pid': 305,
  'port': 53}```

so now should raise an error , and at this case we never install coredns in node 

another solution : 
can i check if the node have coredns container or not i have it i raise error otherwise install coredns and create container